core/src/main/java/org/apache/calcite/plan/MaterializedViewSubstitutionVisitor.java

```diff
@@ -180,6 +182,39 @@ public UnifyResult apply(UnifyRuleCall call) {
     }
   }
 
+  /**
+   * Implementation of {@link SubstitutionVisitor.UnifyRule} that matches a
+   * {@link MutableUnion} to a {@link MutableUnion} where the query and target
+   * have the same inputs but might not have the same order.
+   */
+  private static class UnionToUnionRule extends AbstractUnifyRule {
+    public static final UnionToUnionRule INSTANCE = new UnionToUnionRule();
+
+    private UnionToUnionRule() {
+      super(any(MutableUnion.class), any(MutableUnion.class), 0);
+    }
+
+    public UnifyResult apply(UnifyRuleCall call) {
+      final MutableUnion query = (MutableUnion) call.query;
+      final MutableUnion target = (MutableUnion) call.target;
+      List<MutableRel> queryInputs = query.getInputs();
+      List<MutableRel> targetInputs = target.getInputs();
+      if (queryInputs.size() == targetInputs.size()) {
+        while (!queryInputs.isEmpty()) {
```

Comment:
How about using for loop (i .. n) here? I don't think removing elements in query inputs is necessary. Other than that, looks good to me.

Reply:
hmm~ let me think. The process of removing elements might be useful/necessary.
Let's make an example. Say Query and MV are like below:
```
Query:
Union
+--ScanA
+--ScanA
+--ScanB

MV:
+--ScanA
+--ScanB
+--ScanB
```
Obviously the matching should be failed, but if I implement like below:
```
if (queryInputs.size() == targetInputs.size()) {
  for (MutableRel rel： queryInputs) {
    if (targetInputs.indexOf(queryInputs.get(0)) == -1) {
      return null;
    }
  }
}
return true;
```

The matching will succeeded, which is not correct.

THX again for review this :)

Reply:
No, this is what I mean:
```
if (queryInputs.size() == targetInputs.size()) {
  for (MutableRel rel： queryInputs) {
   int index = targetInputs.indexOf(rel);
    if (index == -1) {
      return null;
    } else {
      targetInputs.remove(index);
    }
  }
}
return true;
```

You don't need to remove twice. Does this make sense?

Reply:
Ah~ I got it~ Thanks ~ refine soon.